docs/proposal/20201103-failure-domain.md

```diff
@@ -172,138 +173,207 @@ CSI/CPI will be configured with region/zone information so nodes and PVs will ha
+// FailureDomainHostGroup as the failure domain
+type FailureDomainHostGroup struct {
+  // name of the host group
+  Name string `json:"name"`
+
+  // compute cluster that this hostgroup belongs to
+  // +optional
+  AutoConfigure *bool `json:"autoConfigure,omitempty"`
 }
 
-// ComputeCluster as the failure domain
-type FailureDomainComputeCluster struct {
-    // name of the compute cluster
-    Name string `json:"name"`
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=vspherefailuredomains,scope=Cluster,categories=cluster-api
+
+// VSphereFailureDomain is the Schema for the vspherefailuredomains API
+type VSphereFailureDomain struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
 
-// datacenter that this compute cluster belongs to
-    Datacenter string `json:"datacenter"`
+  Spec VSphereFailureDomainSpec `json:"spec,omitempty"`
 }
 
-// HostGroup as the failure domain
-type FailureDomainHostGroup struct {
-    // name of the host group
-    Name string `json:"name"`
+```
+
+we'll also introduce the concept of `VSphereDeploymentZone` which allows use to reference `VSphereFailureDomains`
+and combine it with `PlacementConstrains`
+
+```go
+// VSphereDeploymentZoneSpec defines the desired state of VSphereDeploymentZone
+type VSphereDeploymentZoneSpec struct {
+
+  // Server is the address of the vSphere endpoint.
+  Server string `json:"server,omitempty"`
+
+  // failureDomain is the name of the VSphereFailureDomain used for this VSphereDeploymentZone
+  FailureDomain string `json:"failureDomain,omitempty"`
 
-    // datacenter that this hostgroup belongs to
-    Datacenter string `json:"datacenter"`
+  // ControlPlane determines if this failure domain is suitable for use by control plane machines.
+  // +optional
+  ControlPlane *bool `json:"controlPlane,omitempty"`
 
-    // compute cluster that this hostgroup belongs to
-    ComputeCluster string `json:"computeCluster"`
+// the placement constraints which is used within this failure domain
+  PlacementConstaint PlacementConstraint `json:"placementConstraint"`
 }
 
 // PlacementConstraint is the context information for VM placements within a failure domain
 type PlacementConstraint struct {
-    // ResourcePool is the name or inventory path of the resource pool in which
-    // the virtual machine is created/located.
-    // +optional
-    ResourcePool string `json:"resourcePool,omitempty"`
-
-    // Datastore is the name or inventory path of the datastore in which the
-    // virtual machine is created/located.
-    Datastore string `json:"datastore"`
+  // ResourcePool is the name or inventory path of the resource pool in which
+  // the virtual machine is created/located.
+  // +optional
+  ResourcePool string `json:"resourcePool,omitempty"`
+
+  // Datastore is the name or inventory path of the datastore in which the
+  // virtual machine is created/located.
+  // +optional
+  Datastore string `json:"datastore,omitempty"`
```

Comment:
So different StoragePolicies for Placement constraints wont be supported?

Reply:
IIf you have multiple datastores compatible one storage policy should be able to spread accross those (not sure for datastore clusters though)